### PR TITLE
Select finally script based on device facts

### PR DESCRIPTION
**Pull request by bobbywatson3**
* Run finally script based on device facts matching
* finally_script and image_name are now instance variables in bootstrap scripts
* fixed bug where config would not be written after EOS software upgrade
* Ignore intermittent error when rebooting EOS
* Updated unit tests to reflect code changes

### CODE REVIEW

File: aeon_ztp/bin/cumulus_bootstrap.py [hunk 1]
```diff
@@ -75,63 +72,75 @@ def cli_parse(cmdargs=None):
...
+        self.os_name = 'cumulus'
+        self.progname = '%s-bootstrap' % self.os_name
+        self.logfile = self.cli_args.logfile
+        self.log = self.setup_logging(logname=self.progname)
+        self.user, self.passwd = self.get_user_and_passwd()
+        self.image_name = None
+        self.finally_script = None
+        self.dev = None
+
+    def setup_logging(self, logname):
```
Comment (acarpi) [inline aeon_ztp/bin/cumulus_bootstrap.py]: for consistent logging formats you might want to factor out this logic into a separate logging_utils module or similar and reuse it.

File: aeon_ztp/bin/cumulus_bootstrap.py [hunk 2]
```diff
@@ -75,63 +72,75 @@ def cli_parse(cmdargs=None):
...
+        log = logging.getLogger(name=logname)
+        log.setLevel(logging.INFO)
+
+        fmt = logging.Formatter(
+            '%(asctime)s:%(levelname)s:{target}:%(message)s'
+            .format(target=self.target))
+
+        handler = logging.FileHandler(self.logfile) if self.logfile else logging.StreamHandler(sys.stdout)
```
Comment (acarpi) [inline aeon_ztp/bin/cumulus_bootstrap.py]: when FileHandler is used you might still want to add StreamHandler(sys.stdout) ?

File: aeon_ztp/bin/eos_bootstrap.py [hunk 3]
```diff
@@ -84,20 +80,25 @@ class EosBootstrap:
     def __init__(self, server, cli_args):
```
Comment (acarpi) [inline aeon_ztp/bin/eos_bootstrap.py]: why old style python classes? Also, I see in the various bootstrap scripts I see a few functions that could be made part of a base class, potential improvement for the future?

Comment (bobbywatson3) [inline aeon_ztp/bin/eos_bootstrap.py]: Thanks @acarpi, I've updated the classes to a newer style, and we've also discussed moving some of the code out of the bootstrap scripts and into aeon-venos, which would get rid of a lot of the duplicate code.

File: aeon_ztp/bin/cumulus_bootstrap.py (hunk 1 above)
Comment (bobbywatson3) [inline aeon_ztp/bin/cumulus_bootstrap.py]: @acarpi Agreed. We would like to move shared code into aeon-venos.

File: aeon_ztp/api/views.py [hunk 4]
```diff
@@ -247,6 +247,14 @@ def _put_device_facts():
         rec.os_version = rqst_data.get('os_version')
         rec.facts = rqst_data.get('facts')
         rec.updated_at = time_now()
+        if rqst_data.get('image_name'):
```
Comment (bobbywatson3) [inline aeon_ztp/api/views.py]: Thanks. Updates added.

File: aeon_ztp/bin/aztp_os_selector.py [hunk 5]
```diff
@@ -166,6 +166,7 @@ def main():
             })
         hw_match = match_hw_model(dev_data, cfg_data)
         sw_match = match_os_version(dev_data, hw_match.data)
+        finally_script = hw_match.data.get('finally', None)
```
Comment (bobbywatson3) [inline aeon_ztp/bin/aztp_os_selector.py]: Fixed.

**Comment by bobbywatson3** [comment]
I am running a new build with the most recent changes. Once I have verified, I will merge.

**Comment by bobbywatson3** [comment]
Tests with physical and virtual devices passed.

**Pull request merged**